File: dataAnalysis/wordsScience/lev.hs
import qualified Data.MemoCombinators as Memo

-- 莱温斯坦距离: 两个字符串之间，由一个转换为另一个所需的最少编辑次数
-- 应用: DNA 分析, 拼写检查, 语音辨识, 抄袭侦测

lev :: Eq a => [a] -> [a] -> Int
lev a b = levM (length a) (length b)
  where levM = memoize lev'
        lev' i j
          -- 有个字符串为空，则直接返回最长的字符串
          | min i j == 0 = max i j
          -- 同时计算三种情况
          | otherwise = minimum
                        [ ( 1 + levM (i - 1) j )
                        , ( 1 + levM i (j - 1) )
                        , ( ind i j + levM (i - 1) (j - 1) ) ]
        ind i j
          -- 两个字符相等，则操作次数不做变化
          | a !! (i - 1) == b !! (j - 1) = 0
          -- 两个字符不相等，操作次数 +1
          | otherwise = 1

memoize = Memo.memo2 (Memo.integral) (Memo.integral)

main = print $ lev "mercury" "sylvester"

File: dataAnalysis/wordsScience/lev.js
function memoize (func) {
  const cache = {}
  return (...args) => {
    if (cache[args]) {
      return cache[args]
    }
    cache[args] = func(...args)
    return cache[args]
  }
}

// 用了 memoize 效果相当明显，之前没加这个会跑好久都算不出结果
const lev = (xs, ys) => {
  const ind = (i, j) => xs[i - 1] === ys[j - 1] ? 0 : 1
  const _lev = (i, j) => {
    if (Math.min(i, j) === 0) {
      return Math.max(i, j)
    }
    return Math.min(1 + levM(i - 1, j), 1 + levM(i, j - 1), ind(i, j) + levM(i - 1, j - 1))
  }
  const levM = memoize(_lev)
  return levM(xs.length, ys.length)
}

console.log(lev('find the lights', 'there are four lights'))
console.log(lev('abc', 'abe'))
console.log(lev('mercury', 'sylvester'))
